xdsl/dialects/func.py

```diff
@@ -69,6 +69,51 @@ def from_region(name: str, input_types: List[Attribute],
         op = FuncOp.build(attributes=attributes, regions=[region])
         return op
 
+    def replace_argument_type(self, arg: int | BlockArgument,
+                              new_type: Attribute):
+        """
+        Replaces a function argument with a new one
+        """
+
+        if isinstance(arg, int):
+            if len(self.body.blocks[0].args) <= arg:
+                raise IndexError("Block {} does not have argument #{}!".format(
+                    self.body.blocks[0], arg))
+            arg = self.body.blocks[0].args[arg]
+
+        arg.typ = new_type
+        self.update_function_type()
+
+    def update_function_type(self):
+        """
+        Update the function_type attribute to reflect changes in the
+        block argument types.
+        """
+        return_op = self.body.blocks[-1].ops[-1]
```

Comment:
Also, you are setting the return type of functions that are declarations. I'm not sure this is intended, as these functions have return types, but no `return`.

Reply:
I added something that yells at you if you are using an external function, and also handles non-external function without a return op. I'd argue though, that a func op not having a proper return is UB and we could do whatever we want^^

I also added the `is_declaration` thingy